File: scripts/algorithms/N/Number of Good Leaf Nodes Pairs/Number of Good Leaf Nodes Pairs.cpp
// Runtime: 45 ms (Top 87.77%) | Memory: 32.20 MB (Top 98.08%)

class Solution {
public:int k,ans=0;
    int countPairs(TreeNode* root, int distance) {
        k=distance;
        cp(root,0);
        return ans;
    }
	
    vector<int> cp(TreeNode* root,int level)
    {   vector<int>v1;
        if(root==NULL){return v1;}
        if(root->left==NULL&&root->right==NULL){v1.push_back(level);return v1;}
     
        v1 =cp(root->left,level+1);
        vector<int>v2 =cp(root->right,level+1);
        int a=v1.size();
        int b=v2.size();
        
        for(int i=0;i<a;i++)
        {for(int j=0;j<b;j++)
        { if(abs(v1[i]+v2[j]-(2*level))<=k){ans++;} }}
        
        
        int l=v2.size();
        while(l--){v1.push_back(v2.back());
		v2.pop_back();}
        return v1;
    }
    
};

File: scripts/algorithms/N/Number of Good Leaf Nodes Pairs/Number of Good Leaf Nodes Pairs.java
// Runtime: 86 ms (Top 15.89%) | Memory: 53.6 MB (Top 51.73%)
class Solution {
    static int res;
    public int countPairs(TreeNode root, int distance) {
        res=0;
        rec(root,distance);
        return res;
    }
    static List<Integer> rec(TreeNode root,int dist){
        if(root==null){
            return new LinkedList<Integer>();
        }
        List<Integer> left=rec(root.left,dist);
        List<Integer> right=rec(root.right,dist);
        if(left.size()==0&&right.size()==0){
            List<Integer> temp=new LinkedList<>();
            temp.add(1);
            return temp;
        }
        for(int i:left){
            for(int j:right){
                if(i+j<=dist){
                    res++;
                }
            }
        }
        List<Integer> temp=new LinkedList<>();
        for(int i:left){
            temp.add(i+1);
        }
        for(int i:right){
            temp.add(i+1);
        }
        return temp;
    }
}